File: contest/lccup-2020-fall-team/2.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
    bool isMagic(vector<int> target)
    {
        int n = target.size();
        vector<int> a;
        for (int i = 2; i <= n; i += 2)
        {
            a.push_back(i);
        }
        for (int i = 1; i <= n; i += 2)
        {
            a.push_back(i);
        }

        int k = n;
        for (int i = 0; i < n; ++i)
        {
            if (a[i] != target[i])
            {
                k = i;
                break;
            }
        }
        if (!k)
        {
            return false;
        }
        for (int i = 0; i < n; ++i)
        {
            a[i] = i + 1;
        }
        vector<int> answer;
        // for (auto i = a.begin(); i != a.end(); ++i)
        //     std::cout << *i << ' ';
        while (!a.empty())
        {
            vector<int> b;
            for (int i = 1; i < (int)a.size(); i += 2)
            {
                b.push_back(a[i]);
            }
            for (int i = 0; i < (int)a.size(); i += 2)
            {
                b.push_back(a[i]);
            }

            // for (auto i = b.begin(); i != b.end(); ++i)
            //     std::cout << *i << ' ';

            a.clear();
            for (int i = 0; i < (int)b.size(); ++i)
            {
                if (i < k)
                {
                    answer.push_back(b[i]);
                }
                else
                {
                    a.push_back(b[i]);
                }
            }
        }
        return target == answer;
    }
};


// Driver Code
int main()
{
    vector<int> target{2, 4, 3, 1, 5};
    Solution s;
    bool res = s.isMagic(target);
    cout << res;
    return 0;
}

File: contest/lccup-2020-fall-team/2.py
# https://leetcode-cn.com/problems/er94lq/


from typing import List

# 这就是暴力解法，我为什么比赛时TLE？题都搞错了，没有关系 K 的取值。
# https://leetcode-cn.com/submissions/detail/109552345/


# @风自西来毛则东 详细推理分析 https://leetcode-cn.com/problems/er94lq/solution/mo-ni-xi-pai-guo-cheng-xiang-xi-fen-xi-li-jie-kde-/


class Solution:
    def isMagic(self, target: List[int]) -> bool:
        n = len(target)

        # 第一次洗牌，下标从 1 开始，偶数位置放前面
        a = [i for i in range(2, n + 1, 2)]
        a.extend([i for i in range(1, n + 1, 2)])

        # 记录第一次洗牌后与 target 开始不匹配的位置，推理可得 k == 公共前缀长
        k = n
        for i in range(n):
            if a[i] != target[i]:
                k = i
                break

        if not k:
            return False

        for i in range(n):
            a[i] = i + 1

        ans = []
        while a:
            # 偶数位置放到前面
            b = [a[i] for i in range(1, len(a), 2)]
            b.extend([a[i] for i in range(0, len(a), 2)])
            a.clear()

            for i in range(len(b)):
                if i < k:
                    ans.append(b[i])
                else:
                    a.append(b[i])

        return target == ans


target = [2, 4, 3, 1, 5]
sol = Solution()
print(sol.isMagic(target))
